lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -13,184 +13,217 @@
-        implements Cache<K, V>
+class EvictableCache<K, V>
+        extends AbstractLoadingCache<K, V>
+        implements LoadingCache<K, V>
 {
-    /**
-     * @apiNote Piggy-back on {@link CacheBuilder} for cache TTL.
-     */
-    public static <K, V> EvictableCache<K, V> buildWith(CacheBuilder<? super K, Object> cacheBuilder)
-    {
-        return new EvictableCache<>(cacheBuilder);
-    }
-
-    // private final Map<K, Future<V>> map = new ConcurrentHashMap<>();
-    private final Cache<K, Future<V>> delegate;
-
-    private final StatsCounter statsCounter = new SimpleStatsCounter();
+    // Invariant: for every (K, token) entry in the tokens map, there is a live
+    // cache entry (token, ?) in dataCache, that, upon eviction, will cause the tokens'
+    // entry to be removed.
+    private final ConcurrentHashMap<K, Token<K>> tokens = new ConcurrentHashMap<>();
+    private final LoadingCache<Token<K>, V> dataCache;
 
-    private EvictableCache(CacheBuilder<? super K, Object> cacheBuilder)
+    EvictableCache(CacheBuilder<? super Token<K>, ? super V> cacheBuilder, CacheLoader<? super K, V> cacheLoader)
     {
-        requireNonNull(cacheBuilder, "cacheBuilder is null");
-        this.delegate = buildUnsafeCache(cacheBuilder);
+        dataCache = buildUnsafeCache(
+                cacheBuilder
+                        .<Token<K>, V>removalListener(removal -> {
+                            Token<K> token = removal.getKey();
+                            verify(token != null, "token is null");
+                            tokens.remove(token.getKey(), token);
+                        }),
+                new TokenCacheLoader<>(cacheLoader));
     }
 
-    @SuppressModernizer // CacheBuilder.build() is forbidden, advising to use this class as a safety-adding wrapper.
-    private static <K, V> Cache<K, V> buildUnsafeCache(CacheBuilder<? super K, ? super V> cacheBuilder)
+    @SuppressModernizer // CacheBuilder.build(CacheLoader) is forbidden, advising to use this class as a safety-adding wrapper.
+    private static <K, V> LoadingCache<K, V> buildUnsafeCache(CacheBuilder<? super K, ? super V> cacheBuilder, CacheLoader<? super K, V> cacheLoader)
     {
-        return cacheBuilder.build();
+        return cacheBuilder.build(cacheLoader);
     }
 
     @CheckForNull
     @Override
     public V getIfPresent(Object key)
     {
-        Future<V> future = delegate.getIfPresent(key);
-        if (future != null && future.isDone()) {
-            statsCounter.recordHits(1);
-            return getDone(future);
+        @SuppressWarnings("SuspiciousMethodCalls") // Object passed to map as key K
+        Token<K> token = tokens.get(key);
+        if (token == null) {
+            return null;
         }
-        statsCounter.recordMisses(1);
-        return null;
+        return dataCache.getIfPresent(key);
     }
 
     @Override
-    public V get(K key, Callable<? extends V> loader)
+    public V get(K key, Callable<? extends V> valueLoader)
```

Comment:
do you want to restrict access to `putXXX` methods?

Reply:
they should throw, isn't this the case?

Reply:
Yes, it is. I checked now locally.